R/add_cellhash_calls.R

```diff
@@ -0,0 +1,127 @@
+  if (!"barcode_id" %in% colnames(altexp_rowdata)){
+    altexp_rowdata <- altexp_rowdata |>
+      tibble::rownames_to_column("barcode_id")
+  }
+
+
+  barcode_rowdata <- altexp_rowdata |>
+    dplyr::left_join(hashsample_table, by = "barcode_id") |>
+    dplyr::select("barcode_id", "sample_id") |>
+    dplyr::distinct()
+
+  if (any(duplicated(barcode_rowdata$barcode_id))){
+    stop("`barcode_ids` in `hashsample_table can not be duplicated.")
+  }
+
+  # test that each barcode has a corresponding sample
+  missing_barcodes <- barcode_rowdata |>
+    dplyr::filter(is.na(.data$sample_id)) |>
+    dplyr::pull(.data$barcode_id)
+  if (length(missing_barcodes)){
+    warning(paste0("The following `barcode_id`s do not have corresponding `sample_id`s:",
+                   paste0(missing_barcodes, collapse = ", ")))
+  }
+
+  rowData(altExp(sce))$barcode_id <- barcode_rowdata$barcode_id
+  rowData(altExp(sce))$sample_id <- barcode_rowdata$sample_id
+
+  return(sce)
+}
+
+#' Add cellhash demultiplexing results using DropletUtils::hashedDrops
+#'
+#' @param sce SingleCellExperiment object.
+#' @param altexp_id The name of the alternative experiment that contains the cellhash data.
+#'   Default value: "cellhash"
+#' @param ... Other arguments to pass to DropletUtils::hashedDrops
+#'
+#'
+#' @return SingleCellExperiment with colData columns containing the demultiplexing calls
+#'
+#' @import SingleCellExperiment
+#' @importFrom rlang .data
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#'   # add cell calls from DropletUtils::hashedDrops()
+#'   add_demux_hashedDrops(sce = sce)
+#' }
+add_demux_hashedDrops <- function(sce, altexp_id = "cellhash", ...){
+  # check that input is a SingleCellExperiment
+  if(!is(sce, "SingleCellExperiment")){
+    stop("sce must be a SingleCellExperiment object")
+  }
+  if(!altexp_id %in% altExpNames(sce)){
+    stop(glue::glue("altexp_id `{altexp_id}` not found as an experiment in sce"))
+  }
+  # get ids to use: if sample ids are present, use those,
+  #  otherwise barcodes or rownames, in that order
+  sample_ids <- rowData(altExp(sce, altexp_id))$sample_id
+  if(is.null(sample_ids)){
+    sample_ids <- rowData(altExp(sce, altexp_id))$barcode_id
+  }
+  if (is.null(sample_ids)){
+    warning("No sample ids are present for demux results, using row names")
+    sample_ids <- rownames(altExp(sce, altexp_id))
+  }
+
+  # calculate cellhash results
+  hash_result <- DropletUtils::hashedDrops(altExp(sce, altexp_id), ...)
+  stopifnot(all.equal(rownames(hash_result), colnames(sce)))
+
+  # extract results for table
+  hashedDrops_bestid = sample_ids[hash_result$Best]
+  hashedDrops_id = ifelse(hash_result$Confident,
+                          hashedDrops_bestid,
+                          NA_character_)
+
+
+  # add table results to sce
+  sce$hashedDrops_id <- hashedDrops_id
+  sce$hashedDrops_bestid <- hashedDrops_bestid
+  sce$hashedDrops_logfc <- hash_result$LogFC
+  sce$hashedDrops_confident <- hash_result$Confident
```

Comment:
In response to your question about putting all of the results from `hashedDrops()` in the `colData` for `altExp`, I think I do like including all the results there and think it would be helpful rather than picking and choosing. I also do think we need to have the final calls in the main `colData` since that is important. What I was going back and forth on is if it would be beneficial to include the `logFC` also in the main `colData`. But I think I've convinced myself that the main result, which is the call (or the column that you created, `hashedDrops_id`) should be in the main `colData`, with the full output from `hashedDrops()` being in the `altExp` `colData` like you had suggested.